src/python/pants/backend/python/util_rules/pex_environment.py

```diff
@@ -233,9 +199,94 @@ def first_python_binary() -> Optional[PythonExecutable]:
             python_setup.interpreter_search_paths(pex_relevant_environment)
         ),
         subprocess_environment_dict=subprocess_env_vars.vars,
+        # TODO: This path normalization is duplicated with `engine_initializer.py`. How can we do
+        #  the normalization only once, via the options system?
+        named_caches_dir=Path(global_options.options.named_caches_dir).resolve(),
         bootstrap_python=first_python_binary(),
     )
 
 
+@dataclass(frozen=True)
+class CompletePexEnvironment:
+    _pex_environment: PexEnvironment
```

Comment:
Consider prefixing `PexEnvironment` with `_`. Iiuc, it's no longer meant to be consumed by other files.

Reply:
Its perfectly fine to be consumed, its just not very useable. That's reflected by the fact the useful methods are stripped. Right now each consumer uses one or the other CompletePexEnvironment variant. Someone may want to use both. They will consume a PexEnvironment and simple call SandboxBoxPexEnvironment.create and / or WorkspacePexEnvironment.create as needed using plain old python code.